File: web/app/utils/assistant-access.ts
type AssistantAccessContext = {
  role?: string
  effectivePermissionKeys?: readonly string[]
  effectivePermissionsResolved?: boolean
}

const ASSISTANT_CONFIG_VIEW_PERMISSIONS = new Set([
  'automation.manage',
  'calendar.manage',
  'meta_ads.manage',
  'omnichannel.agents.manage',
  'core.account.manage',
  'workspace.configuracoes.edit',
])

const ASSISTANT_CONFIG_MANAGE_PERMISSIONS = new Set([
  'automation.manage',
  'core.account.manage',
  'workspace.configuracoes.edit',
])

function hasAssistantPermission(
  context: AssistantAccessContext,
  accepted: ReadonlySet<string>,
): boolean {
  if (context.role === 'platform_admin' || context.role === 'owner') return true
  if (!context.effectivePermissionsResolved) return false
  return (context.effectivePermissionKeys || []).some((permission) => accepted.has(permission))
}

export function canViewAssistantConfiguration(context: AssistantAccessContext): boolean {
  return hasAssistantPermission(context, ASSISTANT_CONFIG_VIEW_PERMISSIONS)
}

export function canManageAssistantConfiguration(context: AssistantAccessContext): boolean {
  return hasAssistantPermission(context, ASSISTANT_CONFIG_MANAGE_PERMISSIONS)
}

File: back/internal/modules/auth/assistant_access.go
package auth

import "strings"

// CanConfigureAssistant centraliza a politica transversal de configuracao do
// Assistente 360. RequireAuthWithAccount deve ter validado membership e
// hidratado Permissions antes desta checagem.
func CanConfigureAssistant(principal Principal) bool {
	if principal.Role == RolePlatformAdmin || principal.Role == RoleOwner {
		return true
	}
	if !principal.PermissionsResolved {
		return false
	}
	for _, permission := range principal.Permissions {
		switch strings.TrimSpace(permission) {
		case "automation.manage",
			"calendar.manage",
			"meta_ads.manage",
			"omnichannel.agents.manage",
			"core.account.manage",
			"workspace.configuracoes.edit":
			return true
		}
	}
	return false
}

// CanManageAssistantConfiguration protege os campos transversais do Assistente
// 360 (persona, modelo, credencial e matriz de todas as surfaces). Uma permissao
// restrita a Calendar, Meta Ads ou Omnichannel nao pode reconfigurar os demais
// dominios da conta.
func CanManageAssistantConfiguration(principal Principal) bool {
	if principal.Role == RolePlatformAdmin || principal.Role == RoleOwner {
		return true
	}
	if !principal.PermissionsResolved {
		return false
	}
	for _, permission := range principal.Permissions {
		switch strings.TrimSpace(permission) {
		case "automation.manage", "core.account.manage", "workspace.configuracoes.edit":
			return true
		}
	}
	return false
}

// CanManageAssistantCredentials restringe mutacoes do cofre compartilhado a
// administradores transversais. Gestores de um modulo podem selecionar e usar
// entradas mascaradas, mas nao criar, rotacionar ou excluir segredos globais.
func CanManageAssistantCredentials(principal Principal) bool {
	return CanManageAssistantConfiguration(principal)
}
